### PR TITLE
AQL Profiling code

### PR DESCRIPTION
**Pull request by graetzer**
Execute explain as `db._explain("FOR x IN _frontend RETURN x", {}, {profile:2})` to test.
Will print per node runtime stats and more

### CODE REVIEW

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/311/

**Comment by Simran-B** [comment]
Looking forward to this feature!

Will the profile numbers 0-4 be changed to something readable later on? 3 and 4 are not explained yet.

There are unrelated documentation changes in the branch in case you didn't notice:
https://github.com/arangodb/arangodb/pull/5165/files#diff-00cff8b294684243467a555a778569f5

**Comment by graetzer** [comment]
No not likely I think I will keep it like that. I will add a new method `db._profileQuery` for end-users. The remaining integer values  stay internal for now.

**Comment by graetzer** [comment]
And yes I think I fixed some unrelated docs

**Comment by dothebart** [comment]
Will you add bytes transfered to the scatter / gather nodes? Since many small documents in one big chunk may not hurt as much as several big ones...

**Comment by graetzer** [comment]
Perhaps later, lets first get this out the door

**Comment by graetzer** [comment]
```Query string:
 FOR x IN test FILTER x.cc < 25 LIMIT 10 RETURN x

Execution plan:
 Id   NodeType                  Calls   Items   Runtime [s]   Comment
  1   SingletonNode                 1       1        0.0000   * ROOT
  2   EnumerateCollectionNode     102     463        0.0012     - FOR x IN test   /* full collection scan */
  3   CalculationNode             102     463        0.0018       - LET #1 = (x.`cc` < 25)   /* simple expression */   /* collections used: x : test */
  4   FilterNode                    1      10        0.0016       - FILTER #1
  5   LimitNode                     2      10        0.0019       - LIMIT 0, 10
  6   ReturnNode                    2      10        0.0016       - RETURN x

Indexes used:
 none

Optimization rules applied:
 none

Runtime Statistics:
 Writes Exec   Writes Ign   Scan Full   Scan Index   Filtered   Exec Time [s]
           0            0         463            0        453        0.00576s

Query Profile:
 Query Stage           Duration [s]
 initializing          0.00002
 parsing               0.00026
 optimizing ast        0.00003
 loading collections   0.00004
 instantiating plan    0.00019
 optimizing plan       0.00126
 executing             0.00391
 finalizing            0.00000
```

**Comment by graetzer** [comment]
@dothebart @Simran-B  Comments on the layout ?

**Comment by Simran-B** [comment]
Maybe add another column for runtime in percent?

File: arangod/Aql/QueryResult.h [hunk 1]
```diff
@@ -59,18 +58,16 @@ struct QueryResult {
   ////////////////////////////////////////////////////////////////////
   /// @brief builds the "extra" attribute values from the result.
   ////////////////////////////////////////////////////////////////////
-  std::shared_ptr<velocypack::Builder> extra() const;
+  //std::shared_ptr<velocypack::Builder> extra() const;
```
Comment (mchacki) [inline arangod/Aql/QueryResult.h]: please remove

File: arangod/Aql/QueryResult.h [hunk 2]
```diff
@@ -59,18 +58,16 @@ struct QueryResult {
...
   int code;
   bool cached;
   std::string details;
   std::unordered_set<std::string> bindParameters;
   std::vector<std::string> collectionNames;
-  std::shared_ptr<arangodb::velocypack::Builder> warnings;
   std::shared_ptr<arangodb::velocypack::Builder> result;
-  std::shared_ptr<arangodb::velocypack::Builder> stats;
-  std::shared_ptr<arangodb::velocypack::Builder> profile;
+  std::shared_ptr<arangodb::velocypack::Builder> extra;
```
Comment (mchacki) [inline arangod/Aql/QueryResult.h]: good, the simpler the better!

File: arangod/RestHandler/RestExplainHandler.cpp [hunk 3]
```diff
@@ -138,26 +138,21 @@ void RestExplainHandler::explainQuery() {
...
-      result.add("stats", stats);
-    }
-  } else {
-    result.add("stats", VPackSlice::noneSlice());
+    result.add("warnings", VPackSlice::emptyArraySlice());
   }
+  if (extras.hasKey("stats")) {
+    result.add("stats", extras.get("stats"));
+  }/* else {
+    result.add("stats", VPackSlice::noneSlice()); // WTF
```
Comment (mchacki) [inline arangod/RestHandler/RestExplainHandler.cpp]: mhm NONE slice is indeed bad.
Please remove comment

**Comment by Simran-B** [comment]
It works with `db._profileQuery()`, which calls `profileQuery()` of the explainer, whereas `db._explain()` calls `explain()` of the explainer. If we want `db._explain(... , ... , { profile: 2 })` to return the profiling information as well, couldn't `profileQuery()` be dropped and `explain()` be extended to optionally include profiling information?
```

127.0.0.1:8529@_system> db._explain
function (query, bindVars, options) {
  if (typeof query === 'object' && typeof query.toAQL === 'function') {
    query = { query: query.toAQL() };
  }

  if (arguments.length > 1) {
    query = { query: query, bindVars: bindVars, options: options };
  }

  require('@arangodb/aql/explainer').explain(query);
}

127.0.0.1:8529@_system> db._profileQuery
function (query, bindVars, options) {
  options = options || {};
  options.profile = 2;
  query = { query: query, bindVars: bindVars, options: options };
  require('@arangodb/aql/explainer').profileQuery(query);
}

```

File: Documentation/Books/AQL/ExecutionAndPerformance/QueryStatistics.md [hunk 4]
```diff
@@ -47,6 +47,9 @@ The meaning of the statistics attributes is as follows:
...
   the top level.
-
-
-
+* *nodes*: _(optional)_ when the query was executed with the option `profile` set to at least *2*,
+  then this value contains runtime statistics per query execution node. This field contains the
+  node id (in `id`), the number of calls to this node `calls` and the number of items returned
+  by this node `items` (Items are the temporary results returned at this stage). You can correlate
+  this statistics with the `plan` returned in `extra`. For a human readable output you can execute 
+  `db.explain(<query>, <bind-vars, {profile:2})` in the arangosh.
```
[the remark this answers is not in the document]
Comment (dothebart) [reply Documentation/Books/AQL/ExecutionAndPerformance/QueryStatistics.md]: it probably could be good to have the actual output of such an explain executed; maybe using sleep and some obvious command. 
Support would need to be added to the example generation scripts.

File: js/common/modules/@arangodb/aql/explainer.js [hunk 5]
```diff
@@ -529,11 +584,41 @@ function processQuery (query, explain) {
...
+        }
+        if (String(n.items).length > maxItemsLen) {
+          maxCallsLen = String(n.items).length;
+        }
+        let l = String(nodes[n.id].runtime.toFixed(3)).length;
+        if (l > maxRuntimeLen) {
+          maxCallsLen = l;
+        }
+      });
+      // by design the runtime is culmulative right now.
```
Comment (Simran-B) [inline js/common/modules/@arangodb/aql/explainer.js]: Typo: cumulative

**Comment by graetzer** [comment]
No we do not want to use `explain` for this at all. ArangoDB's explain is not executing anything right now (you cannot destroy your data by running explain). 

Suddenly adding an option which will trigger it to actually run a query is a dangerous behaviour change.

**Comment by Simran-B** [comment]
@graetzer Yeah, I also came to the conclusion that this has very little to do with explaining queries other than that the explain output looks almost the same. It should be clear now with the documentation fixed.

**Pull request merged**